cloudevents/sdk/http_events.py

```diff
@@ -11,183 +11,148 @@
-                    f"in parameter {fields_refs_name}, {fieldname} "
-                    f"expected type str but found type "
-                    f"{type(fields_refs[fieldname])}."
-                )
-            else:
-                self.optional_attribute_values[f"ce-{field}"] = field
-
-        # structured data is inside json resp['data']
-        self.data = copy.deepcopy(data) if self.isbinary else \
-            copy.deepcopy(data.get('data', {}))
-
-        self.headers = {
-            **self.required_attribute_values,
-            **self.optional_attribute_values
-        }
-
-    def to_request(
+        self._attributes = {k.lower(): v for k, v in attributes.items()}
+        self.data = data
+        required_by_version = {
+            "1.0": v1.Event._ce_required_fields, "0.3": v03.Event._ce_required_fields}
+        if 'specversion' not in self._attributes:
+            self._attributes['specversion'] = "1.0"
+        if 'id' not in self._attributes:
+            self._attributes['id'] = str(uuid.uuid4())
+        if 'time' not in self._attributes:
+            self._attributes['time'] = datetime.datetime.now(
+                datetime.timezone.utc).isoformat()
+
+        if self._attributes['specversion'] not in required_by_version:
+            raise ValueError(
+                f"Invalid specversion: {self._attributes['specversion']}")
+        required_set = required_by_version[self._attributes['specversion']]
+        # There is no good way to default 'source' and 'type', so this checks for those.
+        # required_set = required_by_version.get(attributes.get("specversion"))
+        if not required_set <= self._attributes.keys():
+            raise ValueError(
+                f"Missing required keys: {required_set - attributes.keys()}")
+
+    def to_http(
         self,
-        data_unmarshaller: typing.Callable = lambda x: json.loads(
-            x.read()
-            .decode('utf-8')
-        )
-    ) -> (dict, dict):
+        format: str = converters.TypeStructured,
+        data_marshaller: types.MarshallerType = None
+    ) -> (dict, typing.Union[bytes, str]):
         """
         Returns a tuple of HTTP headers/body dicts representing this cloudevent
 
-        :param data_unmarshaller: callable function used to read the data io
-        object
-        :type data_unmarshaller: typing.Callable
-        :returns: (http_headers: dict, http_body: dict)
+        :param format: constant specifying an encoding format
+        :type format: str
+        :param data_unmarshaller: callable function used to read the data io object
+        :type data_unmarshaller: types.UnmarshallerType
+        :returns: (http_headers: dict, http_body: bytes or str)
         """
-        converter_type = converters.TypeBinary if self.isbinary else \
-            converters.TypeStructured
+        marshaller_by_format = {converters.TypeStructured: lambda x: x,
+                                converters.TypeBinary: json.dumps}
```

Comment:
this instantiation doesn't need to be re-created on every call, therefore this could be moved to a module variable.